crypto-core/src/main/java/com/palantir/crypto/io/DecryptingSeekableInput.java

```diff
@@ -55,37 +57,29 @@ public DecryptingSeekableInput(SeekableInput delegate, SeekableCipher cipher) {
      */
     @Override
     public void seek(long pos) throws IOException {
-        if (pos == getPos()) {
+        if (pos == decryptedStreamPos) {
             // short-circuit if no work to do
             return;
         }
-        // TODO(markelliot) the object allocation here has non-zero cost, for frequent skipping it may be cheaper to eat
-        // and throw away bytes between the current position and a positive next position
-
-        int blockSize = seekableCipher.getBlockSize();
-
-        // If pos is in the first block then seek to 0 and skip pos bytes
-        // else seek to block n - 1 where block n is the block containing the byte at offset pos
-        // in order to initialize the Cipher with the previous encrypted block
-        final long prevBlock;
-        final int bytesToSkip;
-        if (pos < blockSize) {
-            prevBlock = 0;
-            bytesToSkip = (int) pos;
-        } else {
-            prevBlock = pos / blockSize - 1;
-            bytesToSkip = (int) (pos % blockSize + blockSize);
-        }
 
-        long prevBlockOffset = prevBlock * blockSize;
-        Cipher cipher = seekableCipher.seek(prevBlockOffset);
-        delegate.seek(prevBlockOffset);
+        long newBlockIndex = pos / seekableCipher.getBlockSize();
+
+        // if we're at a new block or we tried to move backwards within the block
+        if (newBlockIndex != blockIndex || pos < decryptedStreamPos) {
+            blockIndex = newBlockIndex;
+
+            currentBlockStartRow = blockIndex * seekableCipher.getBlockSize();
 
-        // Need a new cipher stream since seeking the stream and cipher invalidate the cipher stream's buffer
-        decryptedStream = supplier.getInputStream(delegate, cipher);
+            Cipher cipher = seekableCipher.seek(currentBlockStartRow);
+            delegate.seek(currentBlockStartRow);
+
+            currentBlockDecryptionStream = supplier.getInputStream(delegate, cipher);
+        }
 
         // Skip to the byte offset in the block where 'pos' is located
-        ByteStreams.skipFully(decryptedStream, bytesToSkip);
+        int toSkip = (int) (pos - currentBlockStartRow);
```

Comment:
Could use Guava's `Ints.checkedCast(pos - currentBlockStartRow);` to assert that invariant